### PR TITLE
Decode multipart headers as UTF-8.

### PR DESCRIPTION
**Pull request by Tratcher**
#175 
@yishaigalatzer @muratg @GrabYourPitchforks @Eilon 

Please consider for Beta3.

### CODE REVIEW

**Comment by blowdart** [comment]
Please add tests for unicode wide characters, especially invalid surrogate pairs.

**Comment by GrabYourPitchforks** [comment]
:shipit:

File: src/Microsoft.AspNet.WebUtilities/BufferedReadStream.cs [hunk 1]
```diff
@@ -375,6 +367,13 @@ private void ProcessLineChar(StringBuilder builder, ref bool foundCR, ref bool f
             }
         }
 
+        private string DecodeLine(MemoryStream builder, bool foundCRLF)
+        {
+            // Drop the final CRLF, if any
+            var length = foundCRLF ? builder.Length - 2 : builder.Length;
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.WebUtilities/BufferedReadStream.cs]: could there be just CR at the end unix style?

File: src/Microsoft.AspNet.WebUtilities/BufferedReadStream.cs [hunk 2]
```diff
@@ -321,19 +322,15 @@ public string ReadLine(int lengthLimit)
...
-            // Stream ended with no CRLF.
-            return builder.ToString();
+            return DecodeLine(builder, foundCRLF);
         }
 
         public async Task<string> ReadLineAsync(int lengthLimit, CancellationToken cancellationToken)
         {
             CheckDisposed();
-            StringBuilder builder = new StringBuilder();
+            var builder = new MemoryStream(200);
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.WebUtilities/BufferedReadStream.cs]: is the limitation of 200 enforced elsewhere? Any reason not to just construct with lengthlimit + 1?

Comment (Tratcher) [inline src/Microsoft.AspNet.WebUtilities/BufferedReadStream.cs]: 200 isn't a limit, it's the initial buffer size. I expect most normal header lines to fit within this.

File: src/Microsoft.AspNet.WebUtilities/BufferedReadStream.cs (hunk 1 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.WebUtilities/BufferedReadStream.cs]: No, HTTP and MIME are very specific about requiring CRLF.

**Comment by Tratcher** [comment]
Merged.
